src/V3TraceDecl.cpp

```diff
@@ -348,7 +438,48 @@ class TraceDeclVisitor final : public EmitCBaseVisitor {
 
 public:
     // CONSTRUCTORS
-    explicit TraceDeclVisitor(AstNetlist* nodep) { iterate(nodep); }
+    explicit TraceDeclVisitor(AstNetlist* nodep)
+        : m_topScopep{nodep->topScopep()} {
+        FileLine* const flp = nodep->fileline();
+
+        // Iterate modules to build per scope initialization sub functions
+        iterateAndNextConstNull(nodep->modulesp());
+        UASSERT_OBJ(m_subFuncps.empty(), nodep, "Should have been emptied");
+
+        // Build top level trace initialization functions
+        PathAdjustor pathAdjustor{flp, [&](AstNodeStmt* stmtp) { addToTopFunc(stmtp); }};
+        for (const auto& item : m_scopeSubFuncps) {
+            // Adjust name prefix based on path in hierarchy
+            pathAdjustor.adjust(item.first);
+
+            // Call all sub functions for this path
+            for (AstCFunc* const subFuncp : item.second) {
+                AstCCall* const callp = new AstCCall{flp, subFuncp};
+                callp->argTypes("tracep");
+                addToTopFunc(callp);
+            }
+        }
+        pathAdjustor.unwind();
+
+        // Ensure a top function exists, in case there was nothing to trace at all
```

Comment:
Can you please add a test where tracing is on and there's no signals at all to be traced (in any scope), e.g just a "module t; endmodule"?